### PR TITLE
feat(e2e/ci): auto-build and publish to docker hub e2e init images on release

### PR DESCRIPTION
**Pull request by anhductn2001**
Closes: #3209 

## What is the purpose of the change

> Auto-build and publish to docker hub e2e init images on release

## Brief Changelog
 
  - *Auto-publish the above image on any minor (12.X.0) and major (X.0.0) release*

### CODE REVIEW

File: .github/workflows/e2e-init-chain.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+name: Build and Push E2E Init Chain Images
+
+on:
+  push:
+    tags:
+    - 'v12+.[0-20]+.0+' 
+    - 'v[0-20]+.0+.0+'
```
Comment (p0mvn) [inline .github/workflows/e2e-init-chain.yml:7]: Why not this?

https://github.com/osmosis-labs/osmosis/blob/8ff4c7fe52657319f6b3c299c7002f7aa90d2d7b/.github/workflows/docker.yml#L27

File: .github/workflows/e2e-init-chain.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+name: Build and Push E2E Init Chain Images
```
Comment (p0mvn) [inline .github/workflows/e2e-init-chain.yml:1]: Can you please add an explanation doc similar to this:
https://github.com/osmosis-labs/osmosis/blob/8ff4c7fe52657319f6b3c299c7002f7aa90d2d7b/.github/workflows/docker.yml#L1

File: .github/workflows/e2e-init-chain.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+on:
+  push:
+    tags:
+    - 'v12+.[0-20]+.0+' 
+    - 'v[0-20]+.0+.0+' 
+
+env:
+  DOCKER_REPOSITORY: osmolabs/osmosis-e2e-init-chain
```
Comment (p0mvn) [inline .github/workflows/e2e-init-chain.yml:10]: Consider naming the file `docker-e2e.yml` please

This is to make it more intuitive about the similarity to `docker.yml` while still being clear that this is for e2e

**Review by p0mvn — COMMENTED** [review]
Great work so far! Thank you for working on this

File: .github/workflows/e2e-init-chain.yml (hunk 3 above)
Comment (anhductn2001) [reply .github/workflows/e2e-init-chain.yml:10]: OK, I'll fix it right now

File: .github/workflows/e2e-init-chain.yml (hunk 1 above)
Comment (anhductn2001) [reply .github/workflows/e2e-init-chain.yml:7]: In the issue, you suggested : We should auto-publish the above image on any minor (12.X.0) and major (X.0.0) release. So I wrote it like that

Comment (p0mvn) [inline .github/workflows/e2e-init-chain.yml:7]:
```suggestion
    - 'v[0-9]+.[0-9]+.0'
```

test: https://regex101.com/r/bx6RJY/1

**Comment by anhductn2001** [comment]
I just made a change. Can you check it again? thankss

File: .github/workflows/docker-e2e.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
+# This workflow pushes new osmosis-e2e-init-chain docker images on every new tag.
+#
+# On every new `vX.Y.Z` tag the following images are pushed:
```
Comment (p0mvn) [inline .github/workflows/docker-e2e.yml:3]:
```suggestion
# On every new `vX.Y.0` tag the following images are pushed:
```

File: .github/workflows/docker-e2e.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+#
+# osmolabs/osmosis-e2e-init-chain:X.Y.Z    # is pushed
```
Comment (p0mvn) [inline .github/workflows/docker-e2e.yml:5]:
```suggestion
# osmolabs/osmosis-e2e-init-chain:X.Y.0    # is pushed
```

File: .github/workflows/docker-e2e.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# osmolabs/osmosis-e2e-init-chain:X.Y      # is updated to X.Y.Z
```
Comment (p0mvn) [inline .github/workflows/docker-e2e.yml:6]:
```suggestion
# osmolabs/osmosis-e2e-init-chain:X.Y      # is updated to X.Y.0
```

File: .github/workflows/docker-e2e.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# osmolabs/osmosis-e2e-init-chain:X        # is updated to X.Y.Z
```
Comment (p0mvn) [inline .github/workflows/docker-e2e.yml:7]:
```suggestion
# osmolabs/osmosis-e2e-init-chain:X        # is updated to X.Y.0
```

File: .github/workflows/docker-e2e.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+# All the images above have support for linux/amd64 and linux/arm64.
+#
+# Due to QEMU virtualization used to build multi-platform docker images
+# this workflow might take a while to complete.
+name: Build and Push E2E Init Chain Images
+
+on:
+  push:
+    tags:
+    - 'v[0-9]+.[0-9]+.[0-9]+' # ignore rc
```
Comment (p0mvn) [inline .github/workflows/docker-e2e.yml:18]:
```suggestion
    - 'v[0-9]+.[0-9]+.0' # ignore rc
```

Wait the last one should be 0, isn't it? As per my earlier comment with test cases linked

Comment (anhductn2001) [reply .github/workflows/docker-e2e.yml:18]: I think it should be 0 too. Your last comment suggested changing to [ <> ](https://github.com/osmosis-labs/osmosis/blob/8ff4c7fe52657319f6b3c299c7002f7aa90d2d7b/.github/workflows/docker.yml#L27) 
so I did

**Review by p0mvn — APPROVED** [review]

**Comment by p0mvn** [comment]
On another look, I actually don't understand why we would have a separate file for this.

Can we modify the existing file: https://github.com/osmosis-labs/osmosis/blob/main/.github/workflows/push-docker-images.yml

to add an additional job to build and push init images. That conditional job would only run on patch releases. 

@anhductn2001  Please let me know what you think

**Comment by anhductn2001** [comment]
This push when tag is ` 'v[0-9]+.[0-9]+.0' ` instead of https://github.com/osmosis-labs/osmosis/blob/cc607fcd0cdd5600f5011814029ca9ef08093c7d/.github/workflows/push-docker-images.yml#L27 . 
If this problem can be solved, I think they can be put in a single file.

**Comment by p0mvn** [comment]
> This push when tag is ` 'v[0-9]+.[0-9]+.0' ` instead of https://github.com/osmosis-labs/osmosis/blob/cc607fcd0cdd5600f5011814029ca9ef08093c7d/.github/workflows/push-docker-images.yml#L27 . 
> 
> If this problem can be solved, I think they can be put in a single file.

We should be able to run each individual job conditionally.

For example, our new job/jobs would run only when the patch version is 0. That way we can reuse the same logic while running this only for major and minor releases.

Let me know if that makes sense

File: .github/workflows/docker-e2e.yml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+env:
+  DOCKER_REPOSITORY: osmolabs/osmosis-e2e-init-chain
+
+jobs:
+  docker:
+    runs-on: ubuntu-latest
+    steps:
+      - 
+        name: Check out the repo
+        uses: actions/checkout@v2
```
Comment (niccoloraspa) [inline .github/workflows/docker-e2e.yml:29]: Could you update this to `actions/checkout@v3` please

File: .github/workflows/docker-e2e.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        id: meta_e2e_init_image
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{ env.DOCKER_REPOSITORY }}
+          flavor: |
+            latest=false
+          tags: |
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{major}}
```
Comment (niccoloraspa) [inline .github/workflows/docker-e2e.yml:53]: Could you parse tags as in [here](https://github.com/osmosis-labs/osmosis/blob/c19a29d9e5a51baabb08c0efd7cd2b64accda550/.github/workflows/push-docker-images.yml#L62-L73) ?

The `${{ steps.meta_e2e_init_image.outputs.tags }}` is being deprecated

File: .github/workflows/docker-e2e.yml [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        id: build_push_e2e_init_image
+        uses: docker/build-push-action@v2
+        with:
+          file: tests/e2e/initialization/init.Dockerfile
+          context: .
+          push: true
+          platforms: linux/amd64,linux/arm64
+          build-args: |
+            E2E_SCRIPT_NAME=chain
+          tags: ${{ steps.meta_e2e_init_image.outputs.tags }}
```
Comment (niccoloraspa) [inline .github/workflows/docker-e2e.yml:65]: As stated before, replace this with something like https://github.com/osmosis-labs/osmosis/blob/c19a29d9e5a51baabb08c0efd7cd2b64accda550/.github/workflows/push-docker-images.yml#L89-L92

**Review by niccoloraspa — CHANGES_REQUESTED** [review]
Left some notes, needed to update the GitHub Action after https://github.com/osmosis-labs/osmosis/pull/3390.

**Comment by niccoloraspa** [comment]
> On another look, I actually don't understand why we would have a separate file for this.
> 
> Can we modify the existing file: https://github.com/osmosis-labs/osmosis/blob/main/.github/workflows/push-docker-images.yml
> 
> to add an additional job to build and push init images. That conditional job would only run on patch releases.
> 
> @anhductn2001 Please let me know what you think

Agree to put it in the same file. It's much cleaner.
Please put it in a different job:

```yaml
#...
name: Push Docker Images

on:
  push:
    tags:
    - 'v[0-9]+.[0-9]+.[0-9]+' # ignore rc
  
env:
  DOCKER_REPOSITORY: osmolabs/osmosis
  RUNNER_BASE_IMAGE_DISTROLESS: gcr.io/distroless/static-debian11
  RUNNER_BASE_IMAGE_NONROOT: gcr.io/distroless/static-debian11:nonroot
  RUNNER_BASE_IMAGE_ALPINE: alpine:3.16

jobs:
  docker:      # we could probably rename it to `osmosisd` or `osmosisd_images`
      # .... 
      
  
   e2e_images:      # or `e2e`
     if: ${{ github.ref_name != 'v*.*.0' }}    # Don't run it on patch (please check this)
```

**Comment by anhductn2001** [comment]
I just changed it. Can you review it? Many thanks @niccoloraspa  @p0mvn

**Comment by niccoloraspa** [comment]
Have you tested it?

**Comment by p0mvn** [comment]
> I haven't. Can you tell me how to test it?

- Create a personal fork of `osmosis`
- Merge this branch to `main` there. 
- Try pushing a few tags: major, minor and patch update. 
- See if this job triggers
- Link the results here.

Let me know if you have further questions

**Comment by anhductn2001** [comment]
Hii, Link result here:
- Workflow when push with tag v13.7.0: https://github.com/anhductn2001/osmosis/actions/runs/3485495349
- Workflow when push with tag v13.7.1: https://github.com/anhductn2001/osmosis/actions/runs/3485495800
- Result DockerHub: https://hub.docker.com/repository/docker/ducnt87/osmosis-e2e-init-chain

File: .github/workflows/push-docker-images.yml [hunk 12]
```diff
@@ -33,7 +38,7 @@ env:
   RUNNER_BASE_IMAGE_ALPINE: alpine:3.16
 
 jobs:
-  docker:
+  osmosisd_images:
```
Comment (p0mvn) [inline .github/workflows/push-docker-images.yml:41]:
```suggestion
  osmosisd-images:
```

**Review by p0mvn — APPROVED** [review]
Great work

**Comment by p0mvn** [comment]
> Hii, Link result here:
> 
> * Workflow when push with tag v13.7.0: https://github.com/anhductn2001/osmosis/actions/runs/3485495349
> * Workflow when push with tag v13.7.1: https://github.com/anhductn2001/osmosis/actions/runs/3485495800
> * Result DockerHub: https://hub.docker.com/repository/docker/ducnt87/osmosis-e2e-init-chain

Awesome work on the tests

**Comment by anhductn2001** [comment]
I don't know why Test CosmWasm Contracts failed. It seems all github action about it have a problem. Can you check ??

**Review by p0mvn — APPROVED** [review]

**Comment by p0mvn** [comment]
The failing job is unrelated to this PR. I'm going to merge this since all requested changes are addressed. 

Thanks for the great work here

**Pull request merged**